### DISCUSSION TITLE
Integration tests

### DISCUSSION THREAD

**Discussion by CalamarBicefalo** [eclipse/lsp4e#190] [Q&A]
Hello folks, glad to be inaugurating discussions...

I was wondering what's the best way to implement some neat integration tests for our client.

I was hoping to get some test scaffolding at a high level (LSP I/O) where I could do something like this:
```
spyableServer = startMyLSP4EBasedPlugin()

initRequest = spyableServer.requests[0]

// assert it is an init request with the right options...
```

So to reiterate, thinking of having integration tests where the I/O is all about LSP for very high level sanity checks. I have been browsing your test suite trying to get inspiration but couldn't find anything similar, so I was hoping you could send a pointer in the right direction?

Also feel free to close this right away if it violates your issue reporting policies.

**Comment by mickaelistria** [comment]
LSP4E can generate log of messages, which are then streamed into a file. You may want to enable this log (that's setting a preference) and read monitor this log file.
Another solution if you want to ensure your server receives the right data is to add to your server an operation to get the requests if received, and call that operation in your test. Or you may even add extra notifications via your LS, or use some javax.managent or even jfr framework to allow being notified of particular events.
Another solution is to create some kind of "proxy" language-server which adds a layer that enables to monitor messages programatically.
